dartagnan/src/main/java/com/dat3m/dartagnan/wmm/relation/base/RelCrit.java

```diff
@@ -23,58 +23,46 @@ public RelCrit(){
-                for(Event lock : thread.getCache().getEvents(FilterBasic.get(Tag.Linux.RCU_LOCK))){
-                    for(Event unlock : thread.getCache().getEvents(FilterBasic.get(Tag.Linux.RCU_UNLOCK))){
-                        if(lock.getCId() < unlock.getCId()){
-                            maxTupleSet.add(new Tuple(lock, unlock));
-                        }
-                    }
-                }
-            }
-            removeMutuallyExclusiveTuples(maxTupleSet);
+        if(maxTupleSet != null) {
+            return maxTupleSet;
+        }
+        maxTupleSet = new TupleSet();
+        minTupleSet = new TupleSet();
+        FilterUnion filter = FilterUnion.get(FilterBasic.get(Tag.Linux.RCU_LOCK), FilterBasic.get(Tag.Linux.RCU_UNLOCK));
+        for(Thread thread : task.getProgram().getThreads()) {
+            addMatchingTupleSet(thread.getCache().getEvents(filter), Tag.Linux.RCU_LOCK, Tag.Linux.RCU_UNLOCK);
         }
         return maxTupleSet;
     }
 
-    // TODO: Not the most efficient implementation
-    // Let's see if we need to keep a reference to a thread in events for anything else, and then optimize this method
     @Override
     protected BooleanFormula encodeApprox(SolverContext ctx) {
     	BooleanFormulaManager bmgr = ctx.getFormulaManager().getBooleanFormulaManager();
 		BooleanFormula enc = bmgr.makeTrue();
-        for(Thread thread : task.getProgram().getThreads()){
-            for(Event lock : thread.getCache().getEvents(FilterBasic.get(Tag.Linux.RCU_LOCK))){
-                for(Event unlock : thread.getCache().getEvents(FilterBasic.get(Tag.Linux.RCU_UNLOCK))){
-                    if(lock.getCId() < unlock.getCId()){
-                        Tuple tuple = new Tuple(lock, unlock);
-                        if(encodeTupleSet.contains(tuple)){
-                        	BooleanFormula relation = bmgr.and(getExecPair(lock, unlock, ctx));
-                            for(Event otherLock : thread.getCache().getEvents(FilterBasic.get(Tag.Linux.RCU_LOCK))){
-                                if(lock.getCId() < otherLock.getCId() && otherLock.getCId() < unlock.getCId()){
-                                    relation = bmgr.and(relation, bmgr.not(this.getSMTVar(otherLock, unlock, ctx)));
-                                }
-                            }
-                            for(Event otherUnlock : thread.getCache().getEvents(FilterBasic.get(Tag.Linux.RCU_UNLOCK))){
-                                if(lock.getCId() < otherUnlock.getCId() && otherUnlock.getCId() < unlock.getCId()){
-                                    relation = bmgr.and(relation, bmgr.not(this.getSMTVar(lock, otherUnlock, ctx)));
-                                }
-                            }
-                            enc = bmgr.and(enc, bmgr.equivalence(this.getSMTVar(tuple, ctx), relation));
-                        }
-                    }
+        for(Tuple tuple : maxTupleSet) {
```

Comment:
Should be `encodeTupleSet`. But you still need to go over the `maxTupleSet` to check for interferring locks/unlocks.

Reply:
We would still need to extends `encodeTupleSet` over alternatives with intermediates.  Those need to be encoded, as well.

Reply:
I'm not sure what you mean by "alternatives with intermediates".
I think `RelCrit` behaves nicely in that it has no intrinsic meaning to the anarchic semantics and is only relevant for the memory model. So if a tuple is not in the `encodeSet` then no-one cares about the value of that edge in an execution.
This is unlike the case of exclusive events in `RelRMW`, because pairing conditions need to be satisfied due to the anarchic semantics (under e.g. RISC-V an unpaired or wrongly paired SC has to fail).

Reply:
I just looked at the original code and now I get what you mean. Since critical sections can get nested, we use an encoding that works a little different to the usual pairing encodings. In particular, to decide the pairing of an outer pair of lock/unlock, we need to decide the pairing of the inner nested lock/unlocks. 
But then I think your `addMatchingTupleSet` method is incorrect for this case. It seems to me that it will fail in the case of
```
rcu_lock; // 1
rcu_lock; // 2
rcu_unlock; // 3
rcu_unlock; // 4
```
I suspect that the pair (1, 4) will not be found since there are intervening operations.

Reply:
You were right: `rcu-rscs` must not behave like `rmw`.  The algorithm should now result in {(1,4),(2,3)}.

There are still imprecisions;  In the following program, the new iteration yields the additional edge (1,6) in `maxTupleSet`.
```
rcu_lock; // 1
if x
    rcu_unlock; // 2
else
    rcu_unlock; // 3
if y
    rcu_lock; // 4
else
    rcu_lock; // 5
rcu_unlock; // 6
```
Assume that (1,6) was the only tuple added to `encodeTupleSet`.  All others have to be encoded, as well, since the encoding is
```
(= rcu-rscs(1,6)
  (and
    exec(1)
    exec(6)
    (not rcu-rscs(1,2))
    (not rcu-rscs(1,3))
    (not rcu-rscs(4,6))
    (not rcu-rscs(5,6))
  )
)
```

Reply:
Yes, it depends on how smart you do the static matching and the encoding.
You can establish that that the 4 cases you have listed there are actually valid `minTuples`.
Now you either encode `rcu-rscs(1,2) <=> exec(1) /\ exec(2)` (which enlarges your `encodeSet`) and then use `rcu-rscs(1,2)` or you directly inline `exec(1) /\ exec(2)` consistently.
I would go for whatever is easier to implement.
The difference in the encodings is marginal and won't be noticable in runtime at all.

Reply:
Inlining would have to be recursive.  I implemented the artificial enlarging.